plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.crypto.SecretKey;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class FileSystemExchangeSink
+        implements ExchangeSink
+{
+    public static final String COMMITTED_MARKER_FILE_NAME = "committed";
+    public static final String DATA_FILE_SUFFIX = ".data";
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FileSystemExchangeSink.class).instanceSize();
+    private static final int INTEGER_INSTANCE_SIZE = ClassLayout.parseClass(Integer.class).instanceSize();
+
+    private final FileSystemExchangeStorage exchangeStorage;
+    private final URI outputDirectory;
+    private final int outputPartitionCount;
+    private final int numBuffers;
+    private final Optional<SecretKey> secretKey;
+    private final int writeBufferSize;
+
+    private final Map<Integer, ExchangeStorageWriter> writers = new ConcurrentHashMap<>();
+    private final Map<Integer, SliceOutput> pendingBuffers = new ConcurrentHashMap<>();
+    private final LinkedBlockingQueue<SliceOutput> freeBuffers = new LinkedBlockingQueue<>();
+    private final AtomicReference<Throwable> failure = new AtomicReference<>();
+    private volatile boolean finished;
+    @GuardedBy("this")
+    private CompletableFuture<Void> blockedFuture = new CompletableFuture<>();
+
+    public FileSystemExchangeSink(FileSystemExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.numBuffers = outputPartitionCount * 2; // double buffering to overlap computation and I/O
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        this.writeBufferSize = exchangeStorage.getWriteBufferSizeInBytes();
+
+        for (int i = 0; i < numBuffers; ++i) {
+            freeBuffers.add(Slices.allocate(writeBufferSize).getOutput());
+        }
+    }
+
+    // This return result is more like a best-effort hint. The implementation is not thread-safe,
```

Comment:
`This return result is more like a best-effort hint` -> `The future returned by isBlocked should be considered a best-effort hint.`